Add initial support for GridFS.

otrho opened the pull request:
This is pretty basic at this stage, providing a main function to fetch
an entire file by ObjectID, and another by file name.

In the future we may add support for streaming chunks of the requested
file lazily and other metadata related functionality.

otrho committed:
Add initial support for GridFS.

This is pretty basic at this stage, providing a main function to fetch
an entire file by ObjectID, and another by file name.

In the future we may add support for streaming chunks of the requested
file lazily and other metadata related functionality.

WebFreak001 commented on source/kaleidic/mongo_gridfs.d:
```diff
@@ -0,0 +1,62 @@
+//         gridfsBucketName - full path to the GridFS parent of collections .files and .chunks.
+//         id               - the file object ID, i.e., it's "_id" value.
+// RETURN: The entire file contents in an array.
+// THROWS: File not found, or on corrupt properties.
```
This should be adopted to [ddoc](https://dlang.org/spec/ddoc.html) (only slight changes needed) so the documentation is available in editors and in documentation generators.

otrho replied:
I read the ddoc page and didn't really follow it, TBH.  I've just switched to `/++ ... +/`, hopefully that's what you meant?

WebFreak001 replied:
yeah either `///` (single line), `/** ... */` (multi line) or `/++ ... +/` (nested multi line) means ddoc.

Just there are also standard names for the sections. In this example this is nearly ddoc, you would just change a few things: rename `IN`, `RETURN`, `THROWS` to `Params`, `Returns`, `Throws` and for the parameters instead of `-` after the argument name you use `=` + I think you need to terminate the line after the `Params:` header.

WebFreak001 commented on source/kaleidic/mongo_gridfs.d:
```diff
@@ -0,0 +1,62 @@
+//         gridfsBucketName - full path to the GridFS parent of collections .files and .chunks.
+//         name             - file name
+// RETURN: The entire file contents in an array.
+// THROWS: File not found, or if more than 1 file match name, or on corrupt properties.
```
you can use `/// ditto` to copy the documentation from the last function (and add the `name` parameter to the last function's documentation as well)

WebFreak001 commented on source/kaleidic/mongo_gridfs.d:
```diff
@@ -0,0 +1,62 @@
+    }
+    if (reply.documents.length != 1) {
+        throw new Exception("Requested file not found.");
+    }
```
this code is doing a second query with `gridfsReadFile` (first looking up the document by name, then looking it up again by its id, but we already have the document in memory and could use it)

so the code after this could be made into a separate method that is used in both read file functions.

WebFreak001 commented on source/kaleidic/mongo_gridfs.d:
```diff
@@ -0,0 +1,62 @@
+    // Loop for each chunk and append to our buffer.
+    int numChunks = cast(int)(fileSize / chunkSize + 1);
+    for (int chunkIdx = 0; chunkIdx < numChunks; chunkIdx++) {
+        reply = mongo.query(gridfsBucketName ~ ".chunks", 0, 1,
```
The compiler may not optimize the `gridfsBucketName ~ ".chunks"` into a single allocation, putting it outside the loop can avoid unnecessary allocations.

WebFreak001 commented on source/kaleidic/mongo_gridfs.d:
```diff
@@ -0,0 +1,62 @@
+    contents.reserve(fileSize);
+
+    // Loop for each chunk and append to our buffer.
+    int numChunks = cast(int)(fileSize / chunkSize + 1);
```
this could be off-by-one when fileSize is a multiple of chunkSize.

You could fix this with integer calculations (`(fileSize-1)/chunkSize+1` and add a check before that fileSize > 0) or with floating point and rounding/ceiling, but that could introduce imprecision with huge numbers.

Alternatively wouldn't it also be possible to query for all the chunks without `n` and iterate over them? This would save MongoDB query calls as we already have a cursor then which we can iterate on multiple times. (though a single reply from the server might contain multiple chunks and be larger than we want it to be in memory, but I think it's possible to limit that and what it's doing automatically might not even be an issue)

otrho replied:
Whoops, nice find, how embarrassing.

I initially wrote code to query all the chunks at once and iterate through them, sending them to the OutBuffer.  But I wasn't absolutely sure that the results would be sorted on `n`.  They probably are, but I wasn't certain, and I don't see a way to enforce the sorting using this interface (though you can with other Mongo drivers)... @adamdruppe ?

And yeah, as you mention, doing a single query to fetch all the chunks would mean the entire file would be in memory twice, which would probably be OK 99% of the time, still could be an issue for very large files.  I have a feeling there might be better ways to do it, but I'm not sure this library has all the functionality we'd need and/or I can't really work it out yet, since my knowledge of Mongo internals is negligible (and TBH I'd rather keep it that way :wink:).

WebFreak001 replied:
actually queries in MongoDB don't put the whole file in (application) memory at once, you get fed the chunks while iterating (the op_reply in this library implements the cursor and performs the lazy fetching using ranges - so you can use it transparently in foreach)

The MongoDB docs say the file IDs are monotonously incrementing, but no mention about the chunk IDs. However you could add a MongoDB-side `sort` to your query.

I'm not sure whether MongoDB will load the full file into memory, that might be the case - what do other drivers do here actually?

otrho replied:
Before a few days ago I'd only ever heard of MongoDB as 'one of those nosql databases'.  I really don't know. :)

WebFreak001 replied:
ok found the spec here, there is other interesting stuff in there so sharing this while reading: https://github.com/mongodb/specifications/blob/master/source/gridfs/gridfs-spec.rst

WebFreak001 replied:
The spec says

> implementers retrieve all chunks with files_id equal to id, sorted in ascending order on “n”.

as well as

> As drivers stream the stored file they MUST check that each chunk received is the next expected chunk (i.e. it has the expected "n" value) and that the data field is of the expected length

and

> when downloading a zero length stored file the driver MUST NOT issue a query against the chunks collection

so I believe they intend us to use a single sorted query here.

WebFreak001 commented on test/gridfs/README.md:
```diff
@@ -0,0 +1,18 @@
+## GridFS Test Notes
```
would make sense to run these tests on the GitHub actions CI as well (adjust ci.sh)

WebFreak001 reviewed (changes requested):
nice change! In a later PR I would suggest adding reading the files as a range of chunks but would want to see what Adam thinks.

adamdruppe commented:
Yeah I agree with most everything webfreak said. A range of chunks is nice too but kinda optional, can always come back to that later.

If you don't have time I can do a bunch of the tweaks myself after the merge too, just lemme know. The target to get this merged is next Wednesday, so no huge rush but I would like to make the next sil release.

otrho committed:
Fix some potential issues flagged by review.

- Use /++ .. +/ doc comments.
- Fix potential off by one error for numChunks.
- Split functions to improve modularity and consolidate a couple of
  Mongo queries.

otrho commented:
Agreed, I'd like this and the msgpack datetimes issue to be in the next SIL if possible.

otrho committed:
Update GridFS test scripts to (hopefully) work with GitHub Actions.

adamdruppe commented:
On Wed, Jun 23, 2021 at 10:08:23PM -0700, Toby Hutton wrote:
> I initially wrote code to query all the chunks at once and iterate through them, sending them to the OutBuffer.  But I wasn't absolutely sure that the results would be sorted on `n`.  They probably are, but I wasn't certain, and I don't see a way to enforce the sorting using this interface (though you can with other Mongo drivers)... @adamdruppe ?

so i think i've said this before but my experience with mongo stops at
hello world. i wrote this file just because the spec looked reasonably
approachable and something without dependencies was necessary, so i put
it together

but i don't know much of anything about how to actually do mongo queries
or what the other drivers do since i've never actually used them.

otrho commented:
OK, it's working as is for now.  I still need to get the SIL interface going, can we merge and tag this with an intent to improve it if we find issues?

adamdruppe commented:
tagged 0.0.10
